src/components/teleporter.js

```diff
@@ -0,0 +1,377 @@
+    this.v0 = new THREE.Vector3();
+    this.parabola = Array.from(new Array(this.rayCurve.numPoints), () => new THREE.Vector3());
+    this.hit = false;
+    this.hitPoint = new THREE.Vector3();
+    this.meshes = [];
+    this.raycaster = new THREE.Raycaster();
+    this.rigWorldPosition = new THREE.Vector3();
+    this.newRigWorldPosition = new THREE.Vector3();
+    this.teleportOriginWorldPosition = new THREE.Vector3();
+
+    this.teleportEventDetail = {
+      oldPosition: this.rigWorldPosition,
+      newPosition: this.newRigWorldPosition,
+      hitPoint: this.hitPoint
+    };
+    this.prevHitHeight = 0;
+    this.direction = new THREE.Vector3();
+    this.hitEntity = this.createHitEntity();
+    this.hitEntity.object3D.visible = false;
+    this.el.sceneEl.appendChild(this.hitEntity);
+    this.queryCollisionEntities();
+  },
+
+  teleport: (function() {
+    const teleportOriginWorldPosition = new THREE.Vector3();
+    const newRigLocalPosition = new THREE.Vector3();
+
+    return function teleport() {
+      const rig = document.querySelector("#player-rig");
+      rig.object3D.getWorldPosition(this.rigWorldPosition);
+      this.newRigWorldPosition.copy(this.hitPoint);
+
+      // Offset the rig such that the teleportOrigin is above the hitPoint
+      const teleportOrigin = document.querySelector("#player-camera");
+      teleportOrigin.object3D.getWorldPosition(teleportOriginWorldPosition);
+      this.newRigWorldPosition.sub(teleportOriginWorldPosition).add(this.rigWorldPosition);
+
+      // Always keep the rig at the same offset off the ground after teleporting
+      this.newRigWorldPosition.y = this.rigWorldPosition.y + this.hitPoint.y - this.prevHitHeight;
+      this.prevHitHeight = this.hitPoint.y;
+
+      newRigLocalPosition.copy(this.newRigWorldPosition);
+      if (rig.object3D.parent) {
+        rig.object3D.parent.worldToLocal(newRigLocalPosition);
+      }
+      rig.setAttribute("position", newRigLocalPosition);
+      rig.object3D.matrixNeedsUpdate = true;
+      teleportOrigin.object3D.matrixNeedsUpdate = true;
+
+      this.teleportEventDetail.oldPosition = this.rigWorldPosition;
+      this.teleportEventDetail.newPosition = this.newRigWorldPosition;
+      this.teleportEventDetail.hitPoint = this.hitPoint;
+      this.el.emit("teleported", this.teleportEventDetail);
+    };
+  })(),
+
+  queryCollisionEntities: function() {
+    this.collisionEntities = [].slice.call(this.el.sceneEl.querySelectorAll(this.data.collisionEntities));
+    this.meshes = getMeshes(this.collisionEntities);
+  },
+
+  tick(t, dt) {
+    const userinput = AFRAME.scenes[0].systems.userinput;
+    const { start, confirm, speed } = this.data;
+    const object3D = this.el.object3D;
+
+    if (!this.isTeleporting && userinput.get(start)) {
+      this.isTeleporting = true;
+      this.timeTeleporting = 0;
+      this.hit = false;
+      this.rayCurve.mesh.visible = true;
+      this.rayCurve.mesh.updateMatrixWorld();
```

Comment:
Realize this is probably copypaste from teleport controls, but unclear why this needs to get updated